framework/src/main/java/org/checkerframework/framework/util/QualifierKindHierarchy.java

```diff
@@ -0,0 +1,874 @@
+     */
+    protected void initializeQualifierKindFields(
+            Map<QualifierKind, Set<QualifierKind>> directSuperMap) {
+        for (QualifierKind qualifierKind : nameToQualifierKind.values()) {
+            if (qualifierKind.isPoly) {
+                qualifierKind.superTypes = new TreeSet<>();
+                qualifierKind.superTypes.add(qualifierKind.top);
+            } else {
+                qualifierKind.superTypes = findAllTheSupers(qualifierKind, directSuperMap);
+            }
+        }
+        for (QualifierKind qualifierKind : nameToQualifierKind.values()) {
+            for (QualifierKind top : tops) {
+                if (qualifierKind.isSubtype(top)) {
+                    if (qualifierKind.top == null) {
+                        qualifierKind.top = top;
+                    } else if (qualifierKind.top != top) {
+                        throw new UserError(
+                                "Multiple tops found for qualifier %s. Tops: %s and %s.",
+                                qualifierKind, top, qualifierKind.top);
+                    }
+                }
+            }
+            if (qualifierKind.top == null) {
+                throw new UserError("Qualifier isn't in hierarchy: %s", qualifierKind);
+            }
+        }
+        for (QualifierKind qualifierKind : nameToQualifierKind.values()) {
+            for (QualifierKind bot : bottoms) {
+                if (bot.top != qualifierKind.top) {
+                    continue;
+                }
+                if (qualifierKind.bottom == null) {
+                    qualifierKind.bottom = bot;
+                } else if (qualifierKind.top != bot) {
+                    throw new UserError(
+                            "Multiple bottoms found for qualifier %s. Tops: %s and %s.",
+                            qualifierKind, bot, qualifierKind.bottom);
+                }
+                if (qualifierKind.isPoly) {
+                    bot.superTypes.add(qualifierKind);
+                }
+            }
+        }
+    }
+
+    /**
+     * Returns the set of all qualifiers that are a supertype of {@code qualifierKind}.
+     *
+     * @param qualifierKind the qualifier whose super types should be returned
+     * @param directSuperMap directSuperMap
+     * @return the set of all qualifiers that are a supertype of {@code qualifierKind}
+     * @throws UserError if there is a cycle in the hierarchy
+     */
+    private Set<QualifierKind> findAllTheSupers(
+            QualifierKind qualifierKind, Map<QualifierKind, Set<QualifierKind>> directSuperMap) {
+        Queue<QualifierKind> queue = new ArrayDeque<>(directSuperMap.get(qualifierKind));
+        Set<QualifierKind> allSupers = new TreeSet<>(directSuperMap.get(qualifierKind));
+        while (!queue.isEmpty()) {
+            QualifierKind superQual = queue.remove();
+            if (superQual == qualifierKind) {
+                throw new UserError("Cycle in hierarchy: %s", qualifierKind);
```

Comment:
This logic does not find all cycles.  For example, suppose there is qualifierKind -> A -> B -> A.  This will infinite-loop because superQual will ping-pong between A and B but never be equal to qualifierKind.

Reply:
I added logic to break this loop.